File: backend/app/schemas/board.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class BoardCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=150)
    description: Optional[str] = None


class BoardUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=150)
    description: Optional[str] = None


class BoardMemberAdd(BaseModel):
    user_email_or_username: str
    role: str = Field(default="EDITOR", pattern="^(OWNER|EDITOR|VIEWER)$")


class BoardMemberUpdate(BaseModel):
    role: str = Field(..., pattern="^(OWNER|EDITOR|VIEWER)$")


class BoardMemberOut(BaseModel):
    id: str
    board_id: str
    user_id: str
    role: str  # OWNER, EDITOR, VIEWER
    username: str
    email: str
    avatar_url: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class BoardOut(BaseModel):
    id: str
    workspace_id: str
    name: str
    description: Optional[str] = None
    revision: int
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    role: Optional[str] = "EDITOR"
    objects_count: Optional[int] = 0
    workspace_name: Optional[str] = None
    owner_username: Optional[str] = None

    class Config:
        from_attributes = True

File: frontend/src/types/board.ts
export type ObjectType =
  | "sticky_note"
  | "rectangle"
  | "circle"
  | "text"
  | "connector";

export interface BoardObject {
  id: string;
  board_id: string;
  type: ObjectType;
  x: number;
  y: number;
  width: number;
  height: number;
  rotation?: number;
  z_index?: number;
  color?: string;
  fill?: string;
  stroke?: string;
  stroke_width?: number;
  text?: string;
  properties?: Record<string, any>;
  version: number;
  is_deleted?: boolean;
  created_by?: string;
  last_modified_by?: string;
  created_at?: string;
  updated_at?: string;
}

export interface Board {
  id: string;
  workspace_id: string;
  name: string;
  description?: string | null;
  revision: number;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
  role: "OWNER" | "EDITOR" | "VIEWER";
  objects_count?: number;
  workspace_name?: string;
  owner_username?: string;
}

export interface Workspace {
  id: string;
  name: string;
  owner_id: string;
  created_at: string;
  role: "OWNER" | "ADMIN" | "MEMBER";
  boards_count?: number;
}

export interface BoardMember {
  id: string;
  board_id: string;
  user_id: string;
  role: "OWNER" | "EDITOR" | "VIEWER";
  username: string;
  email: string;
  avatar_url?: string | null;
  created_at: string;
}

export interface PresenceUser {
  user_id: string;
  username: string;
  color: string;
  role: string;
  avatar_url?: string;
  x?: number;
  y?: number;
  last_seen?: number;
}
